File: patterns/_tests/BehavioralPatterns/TemplateMethod/java/unit.test.java
// ============================================================
//  Unit Tests — Template Method: Data Pipeline (Java)
//  Run: javac -cp .:junit-platform-console-standalone.jar unit.test.java && java ...
// ============================================================

import org.junit.jupiter.api.*;
import org.junit.jupiter.api.Nested;
import java.io.*;
import java.util.*;
import static org.junit.jupiter.api.Assertions.*;

public class unit {

    static List<String> capture(Runnable fn) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        PrintStream old = System.out;
        System.setOut(new PrintStream(baos));
        try { fn.run(); } finally { System.setOut(old); }
        return Arrays.asList(baos.toString().split("\\r?\\n"));
    }

    // Minimal concrete pipeline for testing base behaviour
    static DataPipeline makePipeline(List<Map<String, Object>> recs) {
        return new DataPipeline() {
            @Override public String name() { return "TestPipe"; }
            @Override protected void extract()   { records = new ArrayList<>(recs); }
            @Override protected void transform() { }
            public void runValidate() { validate(); }
            public void runLoad()     { load(); }
            public void runReport()   { report(); }
        };
    }

    @Nested
    class ValidateTests {
        @Test void passes_records_with_id_and_name() {
            var p = makePipeline(List.of(Map.of("id","1","name","Alice")));
            capture(() -> ((DataPipeline)p).run());
            // run full to populate; just check errors empty after validate step
            var p2 = makePipeline(List.of(Map.of("id","1","name","Alice")));
            p2.extract(); capture(p2::validate);
            assertTrue(p2.errors.isEmpty());
        }

        @Test void catches_missing_name() {
            var rec = new LinkedHashMap<String, Object>();
            rec.put("id", "1"); rec.put("name", null);
            var p = makePipeline(List.of(rec));
            p.extract(); capture(p::validate);
            assertTrue(p.errors.stream().anyMatch(e -> e.contains("name")));
        }

        @Test void catches_missing_id() {
            var rec = new LinkedHashMap<String, Object>();
            rec.put("id", null); rec.put("name", "Alice");
            var p = makePipeline(List.of(rec));
            p.extract(); capture(p::validate);
            assertTrue(p.errors.stream().anyMatch(e -> e.contains("'id'")));
        }

        @Test void output_contains_records_checked() {
            var p = makePipeline(List.of(Map.of("id","1","name","Alice")));
            p.extract();
            var lines = capture(p::validate);
            assertTrue(lines.stream().anyMatch(l -> l.contains("records checked")));
        }
    }

    @Nested
    class ReportHookTests {
        @Test void prints_no_issues_when_clean() {
            var p = makePipeline(List.of());
            p.extract();
            String out = String.join("\n", capture(p::report));
            assertTrue(out.contains("No issues."));
        }

        @Test void prints_warning_when_errors_exist() {
            var p = makePipeline(List.of());
            p.extract();
            p.errors.add("row 0: missing required field 'name'");
            String out = String.join("\n", capture(p::report));
            assertTrue(out.contains("!"));
        }
    }

    @Nested
    class CsvPipelineTests {
        @Test void extract_loads_three_records() {
            var p = new CsvPipeline();
            capture(p::extract);
            assertEquals(3, p.records.size());
        }

        @Test void transform_trims_whitespace() {
            var p = new CsvPipeline();
            capture(p::extract); capture(p::transform);
            var names = p.records.stream().map(r -> String.valueOf(r.get("name"))).toList();
            assertTrue(names.contains("Bob"));
            assertFalse(names.stream().anyMatch(n -> n.contains("  Bob  ")));
        }

        @Test void transform_nulls_empty_string() {
            var p = new CsvPipeline();
            capture(p::extract); capture(p::transform);
            assertNull(p.records.get(2).get("name"));
        }
    }

    @Nested
    class JsonPipelineTests {
        @Test void extract_loads_three_records() {
            var p = new JsonPipeline();
            capture(p::extract);
            assertEquals(3, p.records.size());
        }

        @Test void transform_flattens_nested() {
            var p = new JsonPipeline();
            capture(p::extract); capture(p::transform);
            assertTrue(p.records.get(0).containsKey("meta_role"));
            assertFalse(p.records.get(0).containsKey("meta"));
        }

        @Test void transform_coerces_numeric_string() {
            var p = new JsonPipeline();
            capture(p::extract); capture(p::transform);
            assertEquals(31, p.records.get(0).get("age"));
        }
    }

    @Nested
    class XmlPipelineTests {
        @Test void extract_loads_three_records() {
            var p = new XmlPipeline();
            capture(p::extract);
            assertEquals(3, p.records.size());
        }

        @Test void transform_coerces_int() {
            var p = new XmlPipeline();
            capture(p::extract); capture(p::transform);
            assertEquals(1, p.records.get(0).get("id"));
        }

        @Test void transform_coerces_float() {
            var p = new XmlPipeline();
            capture(p::extract); capture(p::transform);
            assertEquals(9.4, (Double) p.records.get(0).get("score"), 0.001);
        }

        @Test void transform_coerces_bool_true() {
            var p = new XmlPipeline();
            capture(p::extract); capture(p::transform);
            assertEquals(true, p.records.get(0).get("active"));
        }

        @Test void transform_coerces_bool_false() {
            var p = new XmlPipeline();
            capture(p::extract); capture(p::transform);
            assertEquals(false, p.records.get(1).get("active"));
        }
    }

    @Nested
    class RunOrderTests {
        @Test void calls_steps_in_correct_order() {
            List<String> order = new ArrayList<>();
            DataPipeline spy = new DataPipeline() {
                @Override public String name() { return "Spy"; }
                @Override protected void extract()   { order.add("extract"); records = List.of(Map.of("id","1","name","X")); }
                @Override protected void transform() { order.add("transform"); }
                @Override protected void validate()  { order.add("validate"); errors = new ArrayList<>(); }
                @Override protected void load()      { order.add("load"); }
                @Override protected void report()    { order.add("report"); }
            };
            capture(spy::run);
            assertEquals(List.of("extract","transform","validate","load","report"), order);
        }
    }
}

File: patterns/_tests/BehavioralPatterns/TemplateMethod/csharp/unit.test.cs
// ============================================================
//  Unit Tests — Template Method: Data Pipeline (C#)
//  Run: dotnet test
// ============================================================

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Xunit;

public static class CaptureHelper
{
    private static readonly object Lock = new object();

    public static List<string> Capture(Action fn)
    {
        lock (Lock) {
            var sb  = new System.Text.StringBuilder();
            var old = Console.Out;
            Console.SetOut(new StringWriter(sb));
            try { fn(); } finally { Console.SetOut(old); }
            return sb.ToString().Split('\n', StringSplitOptions.RemoveEmptyEntries)
                                .Select(l => l.TrimEnd('\r')).ToList();
        }
    }
}

public class TemplatePipelineUnitTests
{
    static List<string> Capture(Action fn) => CaptureHelper.Capture(fn);

    // Minimal concrete subclass for testing base behaviour
    sealed class SpyPipeline : DataPipeline
    {
        private readonly List<Dictionary<string, object?>> _recs;
        public SpyPipeline(List<Dictionary<string, object?>>? recs = null) =>
            _recs = recs ?? new();
        public override string Name => "TestPipe";
        public override void Extract()   { Records = new(_recs); }
        public override void Transform() { }
        public void RunValidate() { Validate(); }
        public void RunLoad()     { Load(); }
        public void RunReport()   { Report(); }
        public List<string> Errs  => Errors;
    }

    // ── Validate() ───────────────────────────────────────────
    public class ValidateTests
    {
        [Fact] public void Passes_Records_With_Id_And_Name()
        {
            var p = new SpyPipeline(new() { new() { ["id"]="1", ["name"]="Alice" } });
            p.Extract(); Capture(p.RunValidate);
            Assert.Empty(p.Errs);
        }

        [Fact] public void Catches_Missing_Name()
        {
            var p = new SpyPipeline(new() { new() { ["id"]="1", ["name"]=null } });
            p.Extract(); Capture(p.RunValidate);
            Assert.Contains(p.Errs, e => e.Contains("name"));
        }

        [Fact] public void Catches_Missing_Id()
        {
            var p = new SpyPipeline(new() { new() { ["id"]=null, ["name"]="Alice" } });
            p.Extract(); Capture(p.RunValidate);
            Assert.Contains(p.Errs, e => e.Contains("'id'"));
        }

        [Fact] public void Output_Contains_Records_Checked()
        {
            var p = new SpyPipeline(new() { new() { ["id"]="1", ["name"]="Alice" } });
            p.Extract();
            var lines = Capture(p.RunValidate);
            Assert.Contains(lines, l => l.Contains("records checked"));
        }
    }

    // ── Report() hook ────────────────────────────────────────
    public class ReportHookTests
    {
        [Fact] public void Prints_No_Issues_When_Clean()
        {
            var p = new SpyPipeline();
            p.Extract();
            string out_ = string.Join("\n", Capture(p.RunReport));
            Assert.Contains("No issues.", out_);
        }

        [Fact] public void Prints_Warning_When_Errors_Exist()
        {
            var p = new SpyPipeline();
            p.Extract();
            p.Errs.Add("row 0: missing required field 'name'");
            string out_ = string.Join("\n", Capture(p.RunReport));
            Assert.Contains("!", out_);
        }

        sealed class CustomCsv : CsvPipeline {
            protected override void Report() => Console.WriteLine("CUSTOM REPORT");
        }

        [Fact] public void Hook_Can_Be_Overridden()
        {
            string out_ = string.Join("\n", Capture(() => new CustomCsv().Run()));
            Assert.Contains("CUSTOM REPORT", out_);
            Assert.DoesNotContain("No issues.", out_);
        }
    }

    // ── CsvPipeline ──────────────────────────────────────────
    public class CsvTests
    {
        [Fact] public void Extract_Loads_Three_Records()
        {
            var p = new CsvPipeline();
            Capture(p.Extract_Public);
            Assert.Equal(3, p.Records.Count);
        }

        [Fact] public void Transform_Trims_Whitespace()
        {
            var p = new CsvPipeline();
            Capture(p.Extract_Public); Capture(p.Transform_Public);
            var names = p.Records.Select(r => r["name"]?.ToString()).ToList();
            Assert.Contains("Bob", names);
            Assert.DoesNotContain("  Bob  ", names);
        }

        [Fact] public void Transform_Nulls_Empty_String()
        {
            var p = new CsvPipeline();
            Capture(p.Extract_Public); Capture(p.Transform_Public);
            Assert.Null(p.Records[2]["name"]);
        }
    }

    // ── JsonPipeline ─────────────────────────────────────────
    public class JsonTests
    {
        [Fact] public void Extract_Loads_Three_Records()
        {
            var p = new JsonPipeline();
            Capture(p.Extract_Public);
            Assert.Equal(3, p.Records.Count);
        }

        [Fact] public void Transform_Flattens_Nested()
        {
            var p = new JsonPipeline();
            Capture(p.Extract_Public); Capture(p.Transform_Public);
            Assert.True(p.Records[0].ContainsKey("meta_role"));
            Assert.False(p.Records[0].ContainsKey("meta"));
        }

        [Fact] public void Transform_Coerces_Numeric_String()
        {
            var p = new JsonPipeline();
            Capture(p.Extract_Public); Capture(p.Transform_Public);
            Assert.Equal(31, p.Records[0]["age"]);
        }
    }

    // ── XmlPipeline ──────────────────────────────────────────
    public class XmlTests
    {
        [Fact] public void Extract_Loads_Three_Records()
        {
            var p = new XmlPipeline();
            Capture(p.Extract_Public);
            Assert.Equal(3, p.Records.Count);
        }

        [Fact] public void Transform_Coerces_Int()
        {
            var p = new XmlPipeline();
            Capture(p.Extract_Public); Capture(p.Transform_Public);
            Assert.Equal(1, p.Records[0]["id"]);
        }

        [Fact] public void Transform_Coerces_Float()
        {
            var p = new XmlPipeline();
            Capture(p.Extract_Public); Capture(p.Transform_Public);
            Assert.Equal(9.4, (double)p.Records[0]["score"]!, 3);
        }

        [Fact] public void Transform_Coerces_Bool_True()
        {
            var p = new XmlPipeline();
            Capture(p.Extract_Public); Capture(p.Transform_Public);
            Assert.Equal(true, p.Records[0]["active"]);
        }

        [Fact] public void Transform_Coerces_Bool_False()
        {
            var p = new XmlPipeline();
            Capture(p.Extract_Public); Capture(p.Transform_Public);
            Assert.Equal(false, p.Records[1]["active"]);
        }
    }

    // ── Run() step order ─────────────────────────────────────
    public class RunOrderTests
    {
        sealed class OrderSpy : DataPipeline {
            public List<string> Order = new();
            public override string Name => "Spy";
            public override void Extract()           { Order.Add("extract"); Records = new() { new() {["id"]="1",["name"]="X"} }; }
            public override void Transform()         { Order.Add("transform"); }
            protected override void Validate()          { Order.Add("validate"); Errors.Clear(); }
            protected override void Load()              { Order.Add("load"); }
            protected override void Report()            { Order.Add("report"); }
        }

        [Fact] public void Calls_Steps_In_Correct_Order()
        {
            var spy = new OrderSpy();
            Capture(spy.Run);
            Assert.Equal(new[] { "extract","transform","validate","load","report" }, spy.Order);
        }
    }
}

// Public accessor extensions so tests can call protected pipeline methods
public static class PipelineTestExtensions {
    public static void Extract_Public(this CsvPipeline  p) => p.GetType().GetMethod("Extract",   System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance)!.Invoke(p, null);
    public static void Transform_Public(this CsvPipeline p) => p.GetType().GetMethod("Transform", System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance)!.Invoke(p, null);
    public static void Extract_Public(this JsonPipeline  p) => p.GetType().GetMethod("Extract",   System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance)!.Invoke(p, null);
    public static void Transform_Public(this JsonPipeline p) => p.GetType().GetMethod("Transform", System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance)!.Invoke(p, null);
    public static void Extract_Public(this XmlPipeline  p)  => p.GetType().GetMethod("Extract",   System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance)!.Invoke(p, null);
    public static void Transform_Public(this XmlPipeline p) => p.GetType().GetMethod("Transform", System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance)!.Invoke(p, null);
}
